agent/lib/kontena/websocket/client.rb

```diff
@@ -0,0 +1,845 @@
+  def ssl_cert!
+    fail "not connected" unless @socket
+    return nil unless ssl?
+
+    # raises Kontena::Websocket::SSLVerifyError
+    self.ssl_verify_cert! @socket.peer_cert, @socket.peer_cert_chain
+
+    return @socket.peer_cert
+  end
+
+  # Valid once open
+  #
+  # @raise [RuntimeError] not connected
+  # @return [Integer]
+  def http_status
+    with_driver do |driver|
+      driver.status
+    end
+  end
+
+  # Valid once open
+  #
+  # @raise [RuntimeError] not connected
+  # @return [Websocket::Driver::Headers]
+  def http_headers
+    with_driver do |driver|
+      driver.headers
+    end
+  end
+
+  # Read messages from websocket.
+  #
+  # If a block is given, then this loops and yields messages until closed.
+  # Otherwise, returns the next message, or nil if closed.
+  #
+  # @raise [Kontena::Websocket::CloseError]
+  def read(&block)
+    if block
+      read_yield(&block)
+    else
+      read_return
+    end
+  end
+
+  # Send message frame, either text or binary.
+  #
+  # @param message [String, Array<Integer>]
+  # @raise [ArgumentError] invalid type
+  # @raise [RuntimeError] unable to send (socket closed?)
+  def send(message)
+    case message
+    when String
+      with_driver do |driver|
+        fail unless driver.text(message)
+      end
+    when Array
+      with_driver do |driver|
+        fail unless driver.binary(message)
+      end
+    else
+      raise ArgumentError, "Invalid type: #{message.class}"
+    end
+  end
+
+  # Register pong handler.
+  # Called from the #read thread every ping_interval after received pong.
+  # XXX: called with driver @mutex locked
+  #
+  # The ping interval should be longer than the ping timeout.
+  # If new pings are sent before old pings get any response, then the older pings do not yield on pong.
+  #
+  # @yield [delay] received pong
+  # @yieldparam delay [Float] ping-pong delay in seconds
+  def on_pong(&block)
+    @on_pong = block
+  end
+
+  # Send ping message.
+  #
+  # This is intended to be automatically called from #run per @ping_interval.
+  # Calling it from the #run callbacks also works, but calling it from a different thread
+  # while #run is blocked on read() will ignore the @ping_timeout, unless the server happens
+  # to send something else.
+  #
+  # @raise [RuntimeError] not connected
+  def ping
+    with_driver do |driver|
+      # start ping timeout for next read
+      ping_at = pinging!
+
+      debug "pinging at #{ping_at}"
+
+      fail unless driver.ping(ping_at.utc.strftime(PING_STRFTIME)) do
+        # called from read -> @driver.parse with @mutex held!
+        debug "pong for #{ping_at}"
+
+        # resolve ping timeout, unless this pong is late and we already sent a new one
+        if ping_at == @ping_at
+          ping_delay = pinged!
+
+          debug "ping-pong at #{ping_at} in #{ping_delay}s"
+
+          # XXX: defer call without mutex?
+          @on_pong.call(ping_delay) # TODO: also pass ping_at
```

Comment:
Replaced the `@queue` blocks with just `@message_queue`, so now the driver ping callback -> client `on_pong` callback gets called directly, with the driver mutex held... that means that the following fails with a `ThreadError: deadlock; recursive locking` ATM:

```
client.on_pong do
  client.close
end
```

Need to either accept that, or restore some kind of deferred-pong-callback-execution thing.